sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,1118 @@
+    anc_frame_data : List[G3TimestreamMap]
+        List of G3TimestreamMaps saved for each bound frame. This will be
+        populated on bind and should be used to add copies of the anc data to
+        the detector frames. 
+    """
+    def __init__(self, files, book_id, log=None):
+        self.files = files
+        self.times = None
+        self.blocks = {
+            name: HKBlock(name) 
+            for name in ['ACU_broadcast', 'ACU_summary_output', 
+                         'HWPEncoder_freq']
+        }
+        self.anc_frame_data = None
+        self.out_files = []
+        self.book_id = book_id
+
+        if log is None:
+            self.log = logging.getLogger('bookbinder')
+        else:
+            self.log = log
+    
+    def preprocess(self):
+        """
+        Preprocesses HK data and populates the `data` and `times` objects.
+        """
+        self.log.info("Preprocessing HK Data")
+        frame_iter = get_frame_iter(self.files)
+
+        for fr in frame_iter:
+            if fr['hkagg_type'] != 2:
+                continue
+
+            for block in self.blocks.values():
+                block.process_frame(fr)
+
+        for block in self.blocks.values():
+            block.finalize()
+
+    
+    def bind(self, outdir, times, frame_idxs, file_idxs):
+        """
+        Binds ancillary data.
+
+        Params
+        ----------
+        outdir : path
+            Path where files should be written
+        times : np.ndarray
+            Array of timestamps to write to book
+        frame_idxs : np.ndarray
+            Array mapping sample (in times) to output frame index
+        file_idxs : np.ndarray
+            Array mapping output frame idx to output file idx
+        """
+        self.log.info("Binding ancillary data")
+
+        # Handle file writers
+        writer = None
+        cur_file_idx = None
+        out_files = []
+
+        az, el, boresight, corotation = None, None, None, None
+        block = self.blocks['ACU_broadcast']
+        if 'Corrected_Azimuth' in block.data:
+            az = np.interp(times, block.times, block.data['Corrected_Azimuth'])
+            el = np.interp(times, block.times, block.data['Corrected_Elevation'])
+        if 'Corrected_Corotation' in block.data:
+            boresight = np.interp(times, block.times, block.data['Corrected_Boresight'])
+        if 'Corrected_Corotation' in block.data:
+            corotation = np.interp(times, block.times, block.data['Corrected_Corotation'])
+
+        anc_frame_data = []
+        for oframe_idx in np.unique(frame_idxs):
+            if file_idxs[oframe_idx] != cur_file_idx:
+                close_writer(writer)
+                cur_file_idx = file_idxs[oframe_idx]
+                fname = os.path.join(outdir, f'A_ancil_{cur_file_idx:0>3}.g3')
+                out_files.append(fname)
+                writer = core.G3Writer(fname)
+
+            m = frame_idxs == oframe_idx
+            ts = times[m]
```

Comment:
Will we have, as in the original Bookbinder, a flag indicating which timestamps (in ancil) were interpolated? I imagine that would be passed in along with `times` and masked just like here, then propagated into `anc_data` below.

Reply:
Are you asking if we will have a flag-mask for which `bookbinder.times` values are based on smurf data and which are interpolated? I don't think we really need this since the drop rate is very small, especially if we use timestamps from multiple slots to fill in gaps. @kmharrington do you have an opinion on whether we need to include a flag for this?

Reply:
tbh I'm usually in favor of just "save all the masks" just in case. But IDK what level of complication this adds here. 

Is there ever a point where we interpolate the timestamps here and don't also interpolate the signal data? I.e. if the timesample mask would always be 100% overlapped by the signal mask then I'm not worried about it.

Reply:
Yeah—I’m referring to the samples that are missing in the SMuRF file(s) used to make the reference timestamp list. I believe it’s equivalent to the intersect of the `flag_smurfgaps` from the D-files for those wafers (used to make the list), so we wouldn’t be losing info by not including it (I don’t think). It’s more for convenience than anything—in case someone were to only load the A-files.